dds/DCPS/DCPS_Utils.cpp

```diff
@@ -401,13 +401,28 @@ bool repr_to_encoding_kind(DDS::DataRepresentationId_t repr, Encoding::Kind& kin
   return true;
 }
 
-DDS::DataRepresentationIdSeq get_effective_data_rep_qos(DDS::DataRepresentationIdSeq qos) {
-  DDS::DataRepresentationIdSeq qos_ids = qos;
-  if (qos_ids.length() == 0) {
-    qos_ids.length(1);
-    qos_ids[0] = DDS::XCDR_DATA_REPRESENTATION;
+DDS::DataRepresentationIdSeq get_effective_data_rep_qos(const DDS::DataRepresentationIdSeq& qos) {
+  if (qos.length() == 0) {
+    DDS::DataRepresentationIdSeq ids(1);
+    ids.length(1);
+    ids[0] = DDS::XCDR2_DATA_REPRESENTATION;
+    return ids;
+  } else {
+    for (CORBA::ULong i = 0; i < qos.length(); ++i) {
+      Encoding::Kind kind;
+      if (repr_to_encoding_kind(qos[i], kind) && Encoding::KIND_XCDR1 == kind) {
+        // erase XCDR1
+        const CORBA::ULong e = qos.length() - 1;
+        DDS::DataRepresentationIdSeq ids(e);
+        ids.length(e);
+        for (CORBA::ULong j = 0; j < e; ++j) {
+          ids[j] = qos[j < i ? j : (j + 1)];
+        }
+        return ids;
```

Comment:
Is it possible that `qos` contains more than one `XCDR1`?  I guess this wouldn't make a lot of sense but it could happen.

If we want this to be a general purpose "filter" step, it could just accumulate results into a result sequence.